From 831423030038a9820aa9aa1e96fdc087524d1a6b Mon Sep 17 00:00:00 2001
From: Jack Hill <jackhill3103@gmail.com>
Date: Mon, 18 Dec 2023 14:05:48 +0000
Subject: [PATCH 1/2] Fix behaviour when closing KTorrent to system tray

KStatusNotifierIcon hides the QWindow, while GUI hides the main QWidget.
It seems like hiding both of these was causing issues where the QWidget
was considered visible but the QWindow was considered hidden (i.e. clicking
on the tray icon did not close the window when it was visible)

Test-plan:
- Enable KTorrent's system tray icon
- Close the main KTorrent window
- Click the system tray icon twice

Previously:
- First click shows the window
- Second click does nothing

Now:
- First click shows the window
- Second click hides the window
---
 ktorrent/gui.cpp | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/ktorrent/gui.cpp b/ktorrent/gui.cpp
index 9ef76a698..495079a7e 100644
--- a/ktorrent/gui.cpp
+++ b/ktorrent/gui.cpp
@@ -452,11 +452,13 @@ void GUI::loadState(KSharedConfigPtr cfg)
     show_menu_bar_action->setChecked(!menubar_hidden);
 
     bool minimize_to_system_tray = Settings::alwaysMinimizeToSystemTray();
+
+    const auto w = window()->windowHandle();
     if (Settings::showSystemTrayIcon() && minimize_to_system_tray) {
         Out(SYS_GEN | LOG_DEBUG) << "Starting minimized" << endl;
-        hide();
+        w->hide();
     } else {
-        show();
+        w->show();
     }
 
     setCurrentActivity(central->currentActivity());
@@ -478,7 +480,7 @@ void GUI::saveState(KSharedConfigPtr cfg)
 bool GUI::queryClose()
 {
     if (Settings::showSystemTrayIcon() && !qApp->isSavingSession()) {
-        hide();
+        window()->windowHandle()->hide();
         saveState(KSharedConfig::openConfig());
         return false;
     } else {
@@ -502,7 +504,11 @@ void GUI::updateActions()
 
 void GUI::showOrHide()
 {
-    setVisible(!isVisible());
+    const auto w = window()->windowHandle();
+    if (w->isVisible())
+        w->hide();
+    else
+        w->show();
 }
 
 void GUI::configureNotifications()
-- 
GitLab


From bcdc1ebd2ae5a823b2361269c7b6c544407ad089 Mon Sep 17 00:00:00 2001
From: Jack Hill <jackhill3103@gmail.com>
Date: Mon, 18 Dec 2023 14:29:11 +0000
Subject: [PATCH 2/2] Fix group view not being loaded after ktorrent is exited
 when closed to tray

When closing KTorrent to the tray we were hiding the window and then saving
the window state. This meant GroupView was always considered as not visible,
so the group view was not open by default when launching KTorrent again.

Now we save the window state before hiding, and only if the window is actually
visible.

Also there is no need to save the state in GUI::quit since QCoreApplication::quit
now attempts to close in windows in Qt 6, so queryClose is called anyway.
---
 ktorrent/gui.cpp | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/ktorrent/gui.cpp b/ktorrent/gui.cpp
index 495079a7e..3d55070da 100644
--- a/ktorrent/gui.cpp
+++ b/ktorrent/gui.cpp
@@ -479,12 +479,14 @@ void GUI::saveState(KSharedConfigPtr cfg)
 
 bool GUI::queryClose()
 {
+    if (window()->windowHandle()->isVisible()) {
+        saveState(KSharedConfig::openConfig());
+    }
+
     if (Settings::showSystemTrayIcon() && !qApp->isSavingSession()) {
         window()->windowHandle()->hide();
-        saveState(KSharedConfig::openConfig());
         return false;
     } else {
-        saveState(KSharedConfig::openConfig());
         timer.stop();
         QTimer::singleShot(500, qApp, &QCoreApplication::quit);
         return true;
@@ -493,7 +495,6 @@ bool GUI::queryClose()
 
 void GUI::quit()
 {
-    saveState(KSharedConfig::openConfig());
     qApp->quit();
 }
 
-- 
GitLab

